### ISSUE TITLE
🧹 Chore: Why can't `tm2py` just use Emme's Python environment?

### ISSUE THREAD

**Issue by DavidOry** [BayAreaMetro/tm2py#123]
At some point in the development, we decided to create a single Python environment that can accommodate `Lasso` and `tm2py`. This has proven problematic, as the two packages need substantively different Python environments.

Because `Lasso` is only used to create the network, it does not need to have the same Python environment as `tm2py`. It can be run by itself.

But there are other aspects of `tm2py` that need newer versions of the Python spatial libraries than are distributed in Emme. What are they and can we accomplish them some other way to just use the version of Python distributed in Emme?

Progress:

- [ ] Sufficiently defined
- [ ] Approach decided
- [ ] Implemented

Once we determine what these functions are, we can use this space to discuss them. 

fyi @lmz, @gregerhardt

**Comment by i-am-sijia** [comment]
Quick comment: transit_network.py uses `shapely`.
https://github.com/BayAreaMetro/tm2py/blob/c0f9467e47cfff1bb2fb4c49113dea1751fb6ff0/tm2py/components/network/transit/transit_network.py#L14

https://github.com/BayAreaMetro/tm2py/blob/c0f9467e47cfff1bb2fb4c49113dea1751fb6ff0/tm2py/components/network/transit/transit_network.py#L1226-L1243

**Comment by DavidOry** [comment]
I suggest the next time Emme upgrades Python versions, which should be soon given 3.7's age, we should move `tm2py` to only use the Emme distribution. It's likely things like this that are more convenient to do with newer libraries, but can still be accomplished with older libraries.

**Comment by DavidOry** [comment]
We will use the Python distribution with OpenPaths 24.1.

**Comment by i-am-sijia** [comment]
Plans:
1. Create virtual env in OpenPaths shell.
2. Copy the emme.pth file to the venv folder.
3. Install tm2py in the venv in OpenPaths shell. Skip dependencies if they are already included in OpenPaths (e.g., GDAL), remove dependencies if no longer required. Make tm2py installation requirements lightweight.
4. Update tm2py codes to be compatible with OpenPaths dependencies.
5. Run model to make sure everything runs.

**Comment by DavidOry** [comment]
> Plans:
> 
> 1. Create virtual env in OpenPaths shell.
> 2. Copy the emme.pth file to the venv folder.
> 3. Install tm2py in the venv in OpenPaths shell. Skip dependencies if they are already included in OpenPaths (e.g., GDAL), remove dependencies if no longer required. Make tm2py installation requirements lightweight.
> 4. Update tm2py codes to be compatible with OpenPaths dependencies.
> 5. Run model to make sure everything runs.

@i-am-sijia: Plan A would be no virtual environment, and `tm2py` would not have any dependencies that are not in the OpenPaths distribution. Is there anything blocking this exploration?

**Comment by i-am-sijia** [comment]
> [@i-am-sijia](https://github.com/i-am-sijia): Plan A would be no virtual environment, and `tm2py` would not have any dependencies that are not in the OpenPaths distribution. Is there anything blocking this exploration?

Tried Plan A, but the current `tm2py` has dependencies that are not in the OpenPaths Python.

### 1. Dependencies required by the core tm2py components, but not in OpenPaths:

    - `toml` # used in config.py to load config files. alternative is json
    - `pydantic < 2.0` # used in config.py to validate config files. alternative?
    - `requests` # logger.py, used to post msg to slack about model run; also a deep dependency of the documentation libs. alternative?
    - `scipy` # transit_network.py, using scipy non-negative least squares solver. alternative?

For these dependencies, if we install them directly in OpenPaths shell, without creating a virtual env, they will be installed under: `{user_folder}\AppData\Roaming\Python\Python311\site-packages`, because the default OpenPaths env is read-only. I think it's better to create a venv in OpenPaths shell, if we want to install these dependencies.

### 2. Dependencies required by acceptance criteria (which is now included as a component in `tm2py`), but not in OpenPaths:

    - `geopandas`
   
Note that the acceptance criteria notebooks imports `tm2py.acceptance` but is not otherwise dependent on EMME.

### 3. Dependencies required by developers, but not in OpenPaths:
    - [dev-requirements.txt]()

These should not be installed when we update the tm2pt installation, and tm2py version to >=1.0

### 4. Dependencies required for publishing Docs, but not in OpenPaths:
    - [docs/requirements.txt]()

Move these to GitHub Actions?

**Comment by i-am-sijia** [comment]
Punchlist as next steps:

- [ ] Move [tm2py.acceptance](https://github.com/BayAreaMetro/tm2py/tree/develop/tm2py/acceptance) to https://github.com/BayAreaMetro/tm2py-utils, so that `geopandas` is not a requirement to run TM2. @DavidOry, I suppose this will include moving the following:
    - [ ] [tm2pt/notebooks/](https://github.com/BayAreaMetro/tm2py/tree/develop/notebooks): acceptance-xxx.ipynb; acceptance-xxx.twb; on-board-xxx.ipynb; on-board-xxx.twb; create-boardings-xxx.ipynb
    - [ ] [tm2py/examples/](https://github.com/BayAreaMetro/tm2py/tree/develop/examples): acceptance related files
- [ ] In the tm2py OpenPaths EMME virtual env, test install the developer dependencies to confirm those do not have compatibility issues with OpenPaths.
- [ ] Review the make doc GitHub Actions to make sure the doc dependencies are included there **only** and works as expected.

**Comment by DavidOry** [comment]
> Punchlist as next steps:
> 
> * * [ ]  Move [tm2py.acceptance](https://github.com/BayAreaMetro/tm2py/tree/develop/tm2py/acceptance) to https://github.com/BayAreaMetro/tm2py-utils, so that `geopandas` is not a requirement to run TM2. [@DavidOry](https://github.com/DavidOry), I suppose this will include moving the following:
>     
>     * [ ]  [tm2pt/notebooks/](https://github.com/BayAreaMetro/tm2py/tree/develop/notebooks): acceptance-xxx.ipynb; acceptance-xxx.twb; on-board-xxx.ipynb; on-board-xxx.twb; create-boardings-xxx.ipynb* [ ]  [tm2py/examples/](https://github.com/BayAreaMetro/tm2py/tree/develop/examples): acceptance related files[ ]  In the tm2py OpenPaths EMME virtual env, test install the developer dependencies to confirm those do not have compatibility issues with OpenPaths.[ ]  Review the make doc GitHub Actions to make sure the doc dependencies are included there **only** and works as expected.
> 
> fyi [@lachlan-git](https://github.com/lachlan-git)

Yes, please move all acceptance criteria-related infrastructure to the `utils` repo.
